### PR TITLE
PLAT-2539: createSceneAndSceneItems improvements

### PR DESCRIPTION
**Pull request by brentav**
## Summary
This change set introduces several notable improvements to the `createSceneAndSceneItems` helper, as well as a related breaking change:
- Support has been introduced for supplying an `ordinal` value when creating scene items. (`ordinal` is a 0-based index used for defining a consistent ordering amongst sibling scene items.)
- Placeholder scene items will now be created automatically for scene item creation requests where the requested part/revision is missing. Placeholders will not be created if the `failFast` parameter is set to `true`. Records of placeholders created will be returned with the scene item error data.
- You can now supply a `name` on scene item input data for scene creation.
- Breaking change to remove the `createSceneAndSceneItemsEXPERIMENTAL()` method. The batch based scene creation (formerly considered EXPERIMENTAL) has been promoted to default, and there is now only a single helper method named `createSceneAndSceneItems` which takes in a flat list of scene item data. That method incorporates the additional improvements listed above.

## Test Plan
- Test scene creation with small and large scenes, as well as scenes with missing parts/parents
- Test with `ordinal` supplied with scene item request data
- Test ability to supply scene item `name` with request data

## Release Notes
- Breaking Change: The `createSceneAndSceneItemsEXPERIMENTAL()` method has been removed. Use the `createSceneAndSceneItems()` helper method for all scene creation going forward.
- You can now supply an `ordinal` value for scene items. `ordinal` is a 0-based index used for defining a consistent ordering amongst sibling scene items. In the absence of explicitly supplied `ordinal` values, the scene item order within the input data will be observed and honored for scene creation.
- You can now supply a `name` on scene item input data for scene creation.

## Possible Regressions
None expected

## Dependencies
None

### CODE REVIEW

**Comment by brentav** [comment]
Found the source of the code smells: https://github.com/OpenAPITools/openapi-generator/blob/f5903bc391b5dbafe42e4756f8e649e1be8234a9/modules/openapi-generator/src/main/resources/typescript-axios/apiInner.mustache#L199

File: scripts/generate.sh [hunk 1]
```diff
@@ -15,6 +15,7 @@ docker run --rm -v "${PWD}:/local" openapitools/openapi-generator-cli:v5.4.0 gen
 sed -i "" "s/, COLLECTION_FORMATS, /, /" api.ts
 sed -i "" "s/, setApiKeyToObject, /, /" api.ts
 sed -i "" "s/, setBearerAuthToObject, /, /" api.ts
+sed -i "" "s/baseOptions && baseOptions.headers ? baseOptions.headers :/baseOptions.headers ??/" api.ts
```
Comment (mikemccarty-vertex) [inline scripts/generate.sh:18]: 💯

File: client/helpers/scenes.ts [hunk 2]
```diff
@@ -324,7 +319,7 @@ export async function createSceneAndSceneItemsEXPERIMENTAL({
   polling = { intervalMs: PollIntervalMs, maxAttempts: MaxAttempts },
   returnQueued = false,
   verbose,
-}: CreateSceneAndSceneItemsReqEXPERIMENTAL): Promise<CreateSceneAndSceneItemsResEXPERIMENTAL> {
+}: CreateSceneAndSceneItemsReq): Promise<CreateSceneAndSceneItemsResEXPERIMENTAL> {
```
Comment (mikemccarty-vertex) [inline client/helpers/scenes.ts:322]: You're leaving the `Promise`-d type unchanged?

File: client/helpers/scenes.ts [hunk 3]
```diff
@@ -337,8 +332,46 @@ export async function createSceneAndSceneItemsEXPERIMENTAL({
...
   let sceneItemErrors: SceneItemError[] = [];
-  for (let depth = 0; depth < createSceneItemReqs.length; depth++) {
-    const createItemReqs: CreateSceneItemRequest[] = createSceneItemReqs[depth];
+
+  const reqMap: Map<string, CreateSceneItemRequest[]> = new Map();
+  let nextChildren: CreateSceneItemRequest[] = [];
+  reqMap.set('', nextChildren);
+
+  // create parent map and set ordinals based on request order
+  createSceneItemReqs.forEach((req) => {
```
Comment (mikemccarty-vertex) [inline client/helpers/scenes.ts:341]: It looks like this logic appears in two places and could be consolidated into a named function.

File: client/helpers/scenes.ts [hunk 4]
```diff
@@ -337,8 +332,46 @@ export async function createSceneAndSceneItemsEXPERIMENTAL({
...
+    const siblings = reqMap.get(reqParent);
+    if (req.data.attributes.ordinal === undefined) {
+      req.data.attributes.ordinal = siblings?.length;
+    }
+    siblings?.push(req);
+  });
+
+  // sort all scene item requests into depth sorted array of arrays
+  const depthSortedItems: CreateSceneItemRequest[][] = [];
+  while (nextChildren.length) {
```
Comment (danschultz) [inline client/helpers/scenes.ts:358]: I think there's a bug here. The first pass of `nextChildren.length` will yield `0`. Also, generally you do not want rely on type coercion in JS, better to do `while (nextChildren.length > 0)`.

File: client/helpers/scenes.ts [hunk 5]
```diff
@@ -268,6 +245,24 @@ export async function createSceneItems({
...
+  createSceneItemReqs.forEach((req) => {
+    const reqParent =
+      req.data.attributes.parent ??
+      req.data.relationships.parent?.data.id ??
+      '';
+    if (!reqMap.has(reqParent)) {
+      reqMap.set(reqParent, []);
+    }
+    const siblings = reqMap.get(reqParent);
+    if (req.data.attributes.ordinal === undefined) {
```
Comment (danschultz) [inline client/helpers/scenes.ts:260]: It's preferred to not use strict equals (`===`) when checking for emptiness in JS, `ordinal = null` will fail this check. Prefer `req.data.attributes.ordinal == null`, which'll handle both `null` and `undefined`.

**Review by mikemccarty-vertex — APPROVED** [review]
LGTM

File: client/helpers/scenes.ts (hunk 3 above)
Comment (brentav) [reply client/helpers/scenes.ts:201]: The second instance of that logic has been removed.

**Review by danschultz — APPROVED** [review]

File: client/helpers/scenes.ts [hunk 6]
```diff
@@ -174,7 +165,8 @@ export interface QueuedSceneItem {
   readonly res?: Failure | QueuedJob;
 }
 
-const PollPercentage = 10;
+const defaultPolling: Polling = { intervalMs: 200, maxAttempts: 4500 }; // 15 minute timeout for batch completions
```
Comment (MadisonEhlers-Vertex) [inline client/helpers/scenes.ts:168]: This max attempt seems rather large, even for a polling interval. Should we consider a backoff implementation instead?

File: client/helpers/scenes.ts [hunk 7]
```diff
@@ -187,199 +179,319 @@ export async function createSceneAndSceneItems({
...
-    (i) => isQueuedJob(i.res)
-  );
-  const queued = returnQueued ? createRes.queuedSceneItems : [];
-  if (queuedItems.length === 0 || errors.length === createRes.leaves) {
-    return { errors, queued, scene };
+
+  // sort all scene item requests into depth sorted array of arrays
+  const depthSortedItems: CreateSceneItemRequest[][] = [];
+  // fetch list of scene items with no parent (root items)
+  let nextChildren: CreateSceneItemRequest[] = reqMap.get('') || [];
```
Comment (MadisonEhlers-Vertex) [inline client/helpers/scenes.ts:219]: Is `''` the representation of the root?

Will this implementation support multiple root items in the scene? Or only one?

File: client/helpers/scenes.ts [hunk 8]
```diff
@@ -187,199 +179,319 @@ export async function createSceneAndSceneItems({
...
+  reqMap.delete('');
+  while (nextChildren.length > 0) {
```
Comment (MadisonEhlers-Vertex) [inline client/helpers/scenes.ts:221]: 💡 Might not hurt here to add a setTimeout flag on here as a fail safe.

File: client/helpers/scenes.ts [hunk 9]
```diff
@@ -187,199 +179,319 @@ export async function createSceneAndSceneItems({
...
-      allow404: true,
-      limit,
-      polling,
+  let resultQueuedOps: QueuedBatchOps[] = [];
+  let resultBatchErrors: QueuedBatchOps[] = [];
+  let resultItemErrors: SceneItemError[] = [];
+
+  // if we had any scene item requests with invalid parents,
+  // add error entries indicating so.
+  reqMap.forEach((children: CreateSceneItemRequest[]) => {
```
Comment (MadisonEhlers-Vertex) [inline client/helpers/scenes.ts:244]: This is a bit tricky to follow. Why is `reqMap` now a map of items without parents?

File: client/helpers/scenes.ts [hunk 10]
```diff
@@ -187,199 +179,319 @@ export async function createSceneAndSceneItems({
...
+                    [SceneItemSystemMetadata.MissingSuppliedPartRevisionId]:
+                      toMetadataOrUndefined(
+                        item.attributes.source?.suppliedRevisionId
+                      ),
+                  },
+                  source: undefined,
+                },
+                relationships: {
+                  ...item.relationships,
+                  source: undefined,
```
Comment (MadisonEhlers-Vertex) [inline client/helpers/scenes.ts:336]: Why do we need to copy the `item` here? Is it an option to just return it? Why is the `source` set to undefined on a retry?

**Review by MadisonEhlers-Vertex — COMMENTED** [review]
Few questions/nits

**Pull request merged**